src/Microsoft.Identity.Client/Instance/Discovery/InstanceDiscoveryManager.cs

```diff
@@ -95,16 +122,24 @@ public async Task<InstanceDiscoveryMetadataEntry> GetMetadataEntryAsync(string a
                         return entry;
                     }
 
-                    entry = await FetchNetworkMetadataOrFallbackAsync(requestContext, authorityUri, environment).ConfigureAwait(false);
-
-                    if (entry != null)
+                    // If the user provides metadata, don't go to the network
+                    if (_userMetadataProvider != null)
```

Comment:
it's not clear from the code why we would ever get to the user provided metadata, as it's not visible in which cases the static metadata would not return a value (e.g. I think of the static metadata provider as some static fields we have set in the code, however I also remember that we have something that sets these values the first time we fetch the data from somewhere). Wondering if there is somehow that can be made more clear from the code?

Reply:
Ok, I renamed the Static provider as NetworkCache provider and rewrote a bit the logic, hopefully it will be more clear.

Reply:
I also moved more logic into the network provider, which now populates the cache on its own.